src/drivers/esc_rpm/esc_rpm.cpp

```diff
@@ -0,0 +1,242 @@
+}
+
+EscRpm::EscRpm() :
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{
+	param_get(param_find("ESC_RPM_PULSES"), &_pulses_per_revolution);
+
+	for (size_t i = 0; i < MAX_MOTORS; i++) {
+		_motor_data_array[i].timed_out = true;
+	}
+}
+
+void
+EscRpm::start()
+{
+	// Initialize the GPIO rising edge ISRs and callback
+	gpio_init();
+
+	ScheduleOnInterval(100_ms);
+}
+
+void
+EscRpm::gpio_init()
+{
+	irqstate_t state = px4_enter_critical_section();
+
+	/* * * * * * * * * * * * * * * * * *
+	*
+	* Add more GPIO inputs here and define them in board_config.h if you'd like more.
+	*
+	* GPIO_ESC_RPM_1 = PWM CH5
+	* GPIO_ESC_RPM_2 = PWM CH6
+	*
+	*/
+
+	// Configure GPIO
+	px4_arch_configgpio(GPIO_ESC_RPM_1);
+	px4_arch_configgpio(GPIO_ESC_RPM_2);
+
+	// Setup data ready on rising edge
+	px4_arch_gpiosetevent(GPIO_ESC_RPM_1, true, false, true, &EscRpm::isr_callback1, this);
+	px4_arch_gpiosetevent(GPIO_ESC_RPM_2, true, false, true, &EscRpm::isr_callback2, this);
+
+	px4_leave_critical_section(state);
+}
+
+void
+EscRpm::Run()
+{
+	if (should_exit()) {
+		exit_and_cleanup();
+		return;
+	}
+
+	// TESTING: remove when done testing
+	actuator_outputs_s actuators {};
+	actuators.output[0] = 200;
+	_actuator_outputs_pub.publish(actuators);
+
+	// Check if any of the motors have timed out -- this data is shared with ISR so we need to use a critical section
+	// Alternatively we could ping pong buffer, but this should have minimal overhead
+	irqstate_t state = px4_enter_critical_section();
+
+	int motor_rpm[MAX_MOTORS] {};
+
+	uint64_t now = hrt_absolute_time();
+
+	for (size_t i = 0; i < MAX_MOTORS; i++) {
+
+		uint64_t time_since_last_revolution =  now - _motor_data_array[i].start_time;
+
+		if (time_since_last_revolution > MOTOR_TIMEOUT_US) {
+
+			if (!_motor_data_array[i].timed_out) {
+				PX4_INFO("motor %d timed out", i + 1);
+			}
+
+			_motor_data_array[i].timed_out = true;
+			_motor_data_array[i].pulse_count = 0;
+			_motor_data_array[i].rpm = 0;
+		}
+
+		// We make a copy so we can leave the critical section to publish
+		motor_rpm[i] = _motor_data_array[i].rpm;
+	}
+
+	px4_leave_critical_section(state);
+
+	// Publish data for motors
+	esc_rpm_s report {};
+
+	report.timestamp = now;
+	report.esc_rpm[0] = motor_rpm[0];
+	report.esc_rpm[1] = motor_rpm[1];
+	report.esc_rpm[2] = motor_rpm[2];
+	report.esc_rpm[3] = motor_rpm[3];
+
+	_esc_rpm_pub.publish(report);
+}
+
+int
+EscRpm::isr_callback1(int irq, void *context, void *arg)
+{
+	auto obj = static_cast<EscRpm *>(arg);
+	obj->isr_callback_handler(1);
+	return PX4_OK;
+}
+
+int
+EscRpm::isr_callback2(int irq, void *context, void *arg)
+{
+	auto obj = static_cast<EscRpm *>(arg);
+	obj->isr_callback_handler(2);
+	return PX4_OK;
+}
+
+void
+EscRpm::isr_callback_handler(int motor_number)
+{
+	MotorData *motor = &_motor_data_array[motor_number - 1]; // Motors are labelled 1-4 but indexed 0-3
+
+	if (motor->timed_out) {
+		PX4_INFO("ISR: motor: %d timed out", motor_number);
```

Comment:
For clarity, if this message is left in, it should say "_no longer_ timed out" or similar since we just received a pulse